moirae/estimators/online/dual.py

```diff
@@ -0,0 +1,172 @@
+""" Framework for dual estimation of transient vector and A-SOH"""
+from typing import Union, Tuple, Optional, TypedDict, Literal
+from typing_extensions import Self
+
+import numpy as np
+
+from moirae.models.base import InputQuantities, OutputQuantities, GeneralContainer, HealthVariable, CellModel
+from .utils.model import CellModelWrapper, DegradationModelWrapper, convert_vals_model_to_filter, convert_numpy_to_model
+from moirae.estimators.online import OnlineEstimator
+from .filters.base import BaseFilter
+from .filters.distributions import MultivariateRandomDistribution, MultivariateGaussian
+from .filters.kalman.unscented import UnscentedKalmanFilter as UKF
+
+
+class individual_UKF_tuning_inputs(TypedDict):
+    alpha_param: Optional[float]
+    beta_param: Optional[float]
+    kappa_param: Optional[Union[float, Literal['automatic']]]
+
+
+class dual_ukf_tuning_inputs(TypedDict):
+    transient: Optional[individual_UKF_tuning_inputs]
+    asoh: Optional[individual_UKF_tuning_inputs]
+
+
+class DualEstimator(OnlineEstimator):
+    """
+    In dual estimation, the transient vector and A-SOH are estimated by separate filters. This framework generally
+    avoids numerical errors related to the magnitude differences between values pertaining to transient quantities and
```

Comment:
Is the only purpose due to numerical issues? I figured the dual had something more to do with being able to reduce changes in the ASOH estimates by first attempting to explain measurement differences using transients.

Reply:
I do not know for sure, but the tldr for me is "yes." I guess it depends on the implementation/filter being used. Conceptually, the dual estimation framework is always inferior to the joint, as it cannot adequately capture correlations between transient and A-SOH parameters: it takes the posterior from one, assumes it to be "truth", and does what it has to with it. In the case of UKF, it should help with poor numerical conditioning (to be seen), but nothing else really. With other filters, it may have other benefits.

Reply:
Fascinating! More to ponder :)

Reply:
One other useful example is: imagine we find out that UKF is great for updating transient states, but a particle filter works better for the A-SOH. The dual allows us to use them both separately, while the joint requires the same filter for everything.;